src/python/pants/base/deprecated.py

```diff
@@ -62,85 +64,91 @@ def get_deprecated_tense(removal_version: str) -> str:
     """
     if version_string is None:
-        raise MissingSemanticVersionError("The {} must be provided.".format(version_description))
+        raise MissingSemanticVersionError(f"The {version_string} must be provided.")
     if not isinstance(version_string, str):
         raise BadSemanticVersionError(
-            "The {} must be a version string.".format(version_description)
+            f"The {version_description} must be a version string but was {version_string} with "
+            f"type {type(version_description)}."
         )
+
     try:
-        # NB: packaging will see versions like 1.a.0 as 1a0, and are "valid"
-        # We explicitly want our versions to be of the form x.y.z.
         v = Version(version_string)
-        if len(v.base_version.split(".")) != 3:
-            raise BadSemanticVersionError(
-                "The given {} is not a valid version: "
-                "{}".format(version_description, version_string)
-            )
-        if not v.is_prerelease:
-            raise NonDevSemanticVersionError(
-                "The given {} is not a dev version: {}\n"
-                "Features should generally be removed in the first `dev` release "
-                "of a release cycle.".format(version_description, version_string)
-            )
-        return v
     except InvalidVersion as e:
         raise BadSemanticVersionError(
-            "The given {} {} is not a valid version: "
-            "{}".format(version_description, version_string, e)
+            f"The given {version_description} {version_string} is not a valid version: "
+            f"{repr(e)}"
         )
 
+    # NB: packaging.Version will see versions like 1.a.0 as 1a0 and as "valid".
+    # We explicitly want our versions to be of the form x.y.z.
+    if len(v.base_version.split(".")) != 3:
+        raise BadSemanticVersionError(
+            f"The given {version_description} is not a valid version: "
+            f"{version_description}. Expecting the format `x.y.z.dev0`"
+        )
+    if not v.is_prerelease:
+        raise NonDevSemanticVersionError(
+            f"The given {version_description} is not a dev version: {version_string}\n"
+            "Features should generally be removed in the first `dev` release of a release "
+            "cycle."
+        )
+    return v
+
 
-# TODO: propagate `deprecation_start_version` to other methods in this file!
 def warn_or_error(
     removal_version: str,
-    deprecated_entity_description: str,
+    entity: str,
     hint: Optional[str] = None,
-    deprecation_start_version: Optional[str] = None,
+    *,
+    start_version: str | None = None,
     print_warning: bool = True,
 ) -> None:
-    """Check the removal_version against the current pants version.
+    """Check the removal_version against the current Pants version.
+
+    When choosing a removal version, there is a natural tension between the code-base, which benefits
```

Comment:
Taken from `@deprecated`. `warn_or_error` is used more in practice than that decorator.